endaq/calc/shock.py

```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
         - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
           Documentation for the transfer function class used to characterize the
           relative displacement calculation.
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
+          Documentation for the biquad function used to implement the transfer
+          function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
+          Explicit implementations of digital filter coefficients for shock spectra.
     """
-    with warnings.catch_warnings():
-        warnings.simplefilter("ignore", scipy.signal.BadCoefficients)
+    T = utils.sample_spacing(accel)
+    Q = 1./(2.*damp)
 
-        return scipy.signal.TransferFunction(
-            [0, 2 * damp * omega, omega ** 2],
-            [1, 2 * damp * omega, omega ** 2],
-        ).to_discrete(dt=dt)
+    return accel.apply(
+        functools.partial(
+                scipy.signal.lfilter,
+                *_pseudo_velocity_coeffs(omega, Q, T),
+                axis=0,
+                ),
+        raw=True,
+        )
 
 
-def abs_accel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
+def _relative_disp_static_coeffs(omega, Q, T):
     """
-    Calculate the absolute acceleration for a SDOF system.
+    Calculate the coefficients of the Z-domain transfer function for the
+    relative displacement expressed as equivalent static acceleration according
+    to ISO 18431-4
 
-    The "absolute acceleration" follows the transfer function:
-        H(s) = L{x"(t)}(s) / L{y"(t)}(s) = X(s)/Y(s)
+    :param omega: the natural frequency of the system
+    :param Q: the quality factor of the system
+    :param T: the time step in seconds
+    :return: the coefficients of the Z-domain transfer function b, a
+
+    .. seealso::
+
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
+          Explicit implementations of digital filter coefficients for shock spectra.
+    """
+    A = omega*T/(2.*Q)
+    B = omega*T*np.sqrt(1. - 1/(4.*(Q**2.)))
+    C = (T*omega)
+    q = (1./(2.*(Q**2.)) - 1.)/(np.sqrt(1. - 1./(4.*(Q**2.))))
+
+    b = (
+        ((1 - np.exp(-A)*np.cos(B))/Q - q*np.exp(-A)*np.sin(B) - omega*T)/C,
+        (2*np.exp(-A)*np.cos(B)*omega*T - (1 - np.exp(-2.*A))/Q + 2.*q*np.exp(-A)*np.sin(B))/C,
+        (-np.exp(-2.*A)*(omega*T + 1./Q) + np.exp(-A)*np.cos(B)/Q - q*np.exp(-A)*np.sin(B))/C,
+        )
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
+
+
+def relative_disp_static(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

Comment:
abbreviated form:
```suggestion
def rel_displ_static(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
abbreviated form w/ "acceleration":
```suggestion
def rel_displ_static_accel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form:
```suggestion
def relative_displacement_static(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form w/ "acceleration":
```suggestion
def relative_displacement_static_acceleration(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```